L1Trigger/L1TGlobal/src/AXOL1TLCondition.cc

```diff
@@ -249,9 +257,12 @@ const bool l1t::AXOL1TLCondition::evaluateCondition(const int bxEval) const {
   return condResult;
 }
 
-//in order to set model version from config
+//in order to set model version from menu->triggermenuparser->globalproducer->globalboard->here
 void l1t::AXOL1TLCondition::setModelVersion(const std::string modelversionname) {
-  m_AXOL1TLmodelversion = modelversionname;
+  //cases for model version. if version can't be found, then condition returns false
+  if (modelversionname == "v3") {
+    m_AXOL1TLmodelversion = "GTADModel_v3";
+  }
```

Comment:
I plan to discuss this in the DPG meeting Monday: it can be hard coded or automatically set. While hard coding the supported model versions is a bit cumbersome, it avoids the issue that if menu xml is mistakenly updated out of sync with the cms-dist external model dependencies, the emulator will crash with a segfault. Doing it this way allows it to fill the seeds with 0 with a warning that the model version is unsupported via externals. I'm open to suggestions though of course, and plan to discuss this Monday

Reply:
> While hard coding the supported model versions is a bit cumbersome, it avoids the issue that if menu xml is mistakenly updated out of sync with the cms-dist external model dependencies, the emulator will crash with a segfault.

Well.. there are real sw experts here, and I'm not one of them, but this sounds very fragile, and hard-coding names in cpp to cover for it does not sound good either. I guess one way is to improve the [`ModelLoader`](https://github.com/cms-hls4ml/hls4mlEmulatorExtras/blob/main/src/hls4ml/emulator.cc) such that one can check if the model/library is valid before trying to load it. I do agree that the L1T emulator should guarantee that there are no seg-faults inside the HLT due to any of this.

Reply:
I am not sure how to do this myself (definitely no sw expert!) but if someone can suggest how to get model versions from cms-dists to check they are valid that would be a good alternative

Reply:
the way it is currently setup is that it would fill the bit to 0 and print a warning message that the model version is incompatible https://github.com/quinnanm/cmssw/blob/axol1tl_emulator_modelfrommenupr/L1Trigger/L1TGlobal/src/AXOL1TLCondition.cc#L97. So it does not evaluate the condition.

Reply:
> So what exactly would be the desired behavior in case of a mismatch between the menu xml and cmsdist external?

Thinking of HLT, I don't think it should stop the job (seg-fault, exception, or else) [*].

I would say a `LogWarning` would be useful, if not for the fact that the model is loaded (at least) once per event (afaiu), so it would be too many warnings (unless some upper limit on these warnings is added.. or maybe the model is loaded once per run).

[*] Online at HLT, the re-emulated decisions of AXO seeds are currently not used in any way, and this is true also for some other seeds. Iirc (L1T might correct me), in 2022 the re-emulation of MuonShower seeds was incorrect for a while, and this had no impact on HLT outputs. In fact, if there is a disagreement between the L1T decisions in RAW vs those re-emulated inside HLT, I do not think HLT even issues a `LogWarning`. Iiuc, there are L1T plots in the online-DQM which should show the (dis)agreement between the decisions of L1T seeds in firmware-vs-emulation.

Reply:
I see [`ModelLoader::load_model()`](https://github.com/cms-hls4ml/hls4mlEmulatorExtras/blob/main/src/hls4ml/emulator.cc) throws a `std::runtime_error` exception in case it fails to load the shared library, or to find the functions to create and destroy the model. (so naively the "model load" failure should not result a segfault; if the job failure in that case is really a segfault, then something is not even remotely exception safe, and should really be fixed)

Is the `l1t::AXOL1TLCondition::evaluateCondition()` called on every Event? If yes, calling `ModelLoader::load_module()` (that calls `dlopen()` and `dlsym()`) that often sounds potentially expensive. It would be much better to load the model only once (per IOV).

While one could add `try`-`catch` around the call to `model = loader.load_model();`, it would be an anti-pattern for this case  (exceptions are expensive and should not be used for control flow, and frequent caught exceptions make investigations of rare uncaught exceptions difficult).

A better way would be to make the `ModelLoader::load_module()` to communicate by return value (e.g. null `shared_ptr`) in this case) there was an error, so the caller can decide whether it can continue despite of the module loading error. One could even make it choice of the caller whether `ModelLoader::load_module()` would signal the load error via an exception (which should probably be the default?) or a return value.

One could also think more elaborate mechanisms to e.g. be able to list the available models, but those would require more development (and doesn't sound really necessary for this case).

Just curious, how would the mismatch be noticed? From e.g. online DQM?

Reply:
I should note that the model loading only once rather than per event is something that is in development, but is still ongoing. It requires a change to the core eternal model class that is being worked out. However, according to the timing studies this is not expensive or impacting the timing much if at all. Therefore it is something that is in progress but not (as far as I understand) required for initial deployment

Reply:
To be clear, a xml/cms-dist mismatch in the case of a new model being deployed would be unlikely to begin with, but I think it is a good thing to be prepared for just in case. In the current "hard coded" method it would be noticed through the warning message and the fact that no axol1tl conditions pass (all return false).

Reply:
Perhaps we can leave the hardcoded version for now while the core `ModelLoader::load_module()` functionality is changed to enable loading only once and returning a null shared pointer rather than a `std::runtime_error`? This core model class is used by AXOL1TL, CICADA and any other hls4ml-based model, so some care needs to be taken in making any changes

Reply:
@missirol @makortel  apologies, I am only just getting to this. The original idea of this, as requested by L1T O2O experts, the DPG and offline software group was to try and provide multiple layers of fail-safe-ing for HLT so that the way AXOL1TL models are used couldn't (famous last words) cause failures at HLT. The approximate desired set of steps was "Try and load menu mentioned AXOL1TL model library, if that fails, try to load a last known good one, if that fails just return a null/zero AXOL1TL score and we'll deal with the emulator mismatching"

L1T thought this should suffice, but let me ask whether this seems an appropriate course of action to you?

If it is, then it seems the implementation of how that is done should be checked:

The first step in that three step list is now menu read courtesy of new UTM and menu developments. What would be optimal for the second one? Hard-coding it here doesn't seem optimal, would having the slightly more configurable python contain a default be a better option? Having the third one simply skip the condition or output a null score seems fine to me, but is there anything about that should be checked?

> Is the `l1t::AXOL1TLCondition::evaluateCondition()` called on every Event? If yes, calling `ModelLoader::load_module()` (that calls `dlopen()` and `dlsym()`) that often sounds potentially expensive. It would be much better to load the model only once (per IOV).
> 

As an aside, yes. We have requested this, but the AXOL1TL team assures us HLT timing is still met and this development has run into some difficulties, so the O2O problem was deemed higher priority.

> While one could add `try`-`catch` around the call to `model = loader.load_model();`, it would be an anti-pattern for this case (exceptions are expensive and should not be used for control flow, and frequent caught exceptions make investigations of rare uncaught exceptions difficult).
> 
> A better way would be to make the `ModelLoader::load_module()` to communicate by return value (e.g. null `shared_ptr`) in this case) there was an error, so the caller can decide whether it can continue despite of the module loading error. One could even make it choice of the caller whether `ModelLoader::load_module()` would signal the load error via an exception (which should probably be the default?) or a return value.
> 

I can try to make this change, but am perennially short of time to make upgrades to L1T software I would like, so I can't promise any timetable for this.

> Just curious, how would the mismatch be noticed? From e.g. online DQM?

DQM would notice mismatched trigger bits for every AXOL1TL event. HLT itself should be pretty resilient to this kind of mismatch for the moment (it doesn't use the produced score, and AXOL1TL doesn't offer it since the result is created used and destroyed in one action, again, something we have asked about changing). I am not too familiar with the GEN applications, so I could only guess what that means there?

Reply:
> "Try and load menu mentioned AXOL1TL model library, if that fails, try to load a last known good one, if that fails just return a null/zero AXOL1TL score and we'll deal with the emulator mismatching"

My uninformed opinion is that the 2nd step ("try to load a last known good one") is not necessary. Related to this,

> would having the slightly more configurable python contain a default be a better option?

I think the main goals from a HLT point of view are (a) guarantee that the job does not crash, and (b) do not introduce extra parameters in the python configuration of HLT (which HLT then has to maintain).

In short, I would say 'specify the model via the L1T menu, and if it's not available, return `False` for the emulated result (plus one warning per run)'.

Reply:
> > While one could add `try`-`catch` around the call to `model = loader.load_model();`, it would be an anti-pattern for this case (exceptions are expensive and should not be used for control flow, and frequent caught exceptions make investigations of rare uncaught exceptions difficult).
> > A better way would be to make the `ModelLoader::load_module()` to communicate by return value (e.g. null `shared_ptr`) in this case) there was an error, so the caller can decide whether it can continue despite of the module loading error. One could even make it choice of the caller whether `ModelLoader::load_module()` would signal the load error via an exception (which should probably be the default?) or a return value.
> 
> I can try to make this change, but am perennially short of time to make upgrades to L1T software I would like, so I can't promise any timetable for this.

After some more thought, and given the earlier comment that such mismatch should be rare ("exceptional") https://github.com/cms-sw/cmssw/pull/44054#discussion_r1501133458, and that the errors in [`ModelLoader::load_model()`](https://github.com/cms-hls4ml/hls4mlEmulatorExtras/blob/main/src/hls4ml/emulator.cc) contain useful information about the error conditions (beyond just "it failed"), I think the exception would be a better way to communicate the error to the caller of `load_model()` (compared to a return value).

In the HLT case the caller would then catch the exception, maybe log the exact error message, and "fill the bit to 0".

In the offline processing case (I assume this could would get run as part of offline processing as well?), I'd think we'd want the exception to propagate to the framework in order to get an early signal of the mismatch. So maybe make this swallow vs. re-throw the exception behavior configurable?

Two more suggestions for [`ModelLoader::load_model()`](https://github.com/cms-hls4ml/hls4mlEmulatorExtras/blob/main/src/hls4ml/emulator.cc) for longer term (maybe these should go into some issue?)
* It would be better to use a specific exception type instead of `std::runtime_error`, so that the caller-side `catch` would catch only these specific errors from `load_model()`
* It would be better to include the message from the `dlerror()` in the exception message in all cases

Reply:
> > "Try and load menu mentioned AXOL1TL model library, if that fails, try to load a last known good one, if that fails just return a null/zero AXOL1TL score and we'll deal with the emulator mismatching"
> 
> My uninformed opinion is that the 2nd step ("try to load a last known good one") is not necessary. Related to this,
> 
> > would having the slightly more configurable python contain a default be a better option?
> 
> I think the main goals from a HLT point of view are (a) guarantee that the job does not crash, and (b) do not introduce extra parameters in the python configuration of HLT (which HLT then has to maintain).
> 
> In short, I would say 'specify the model via the L1T menu, and if it's not available, return `False` for the emulated result (plus one warning per run)'.

Okay, I think's that's relatively simple, contingent on figuring out a good way to communicate and not crash from the current emulation technique handle. @quinnanm & @artlbv Would it be possible to make these changes if I look at the emulator class for the models?

Reply:
> Then we are okay with some limited version of exception catching here for fail-safe versus fail-out behavior?

I didn't understand what exactly you meant with this. I'm guessing something along
```cpp
try {
  model = loader.load_model();
} catch(std::runtime_error& e) {
  // maybe log e.what()?
  return false;
}
```

It's probably ok for the immediate needs.

But for future offline (mainly simulation?) needs I think it would be good to have a way to fail the jobs early in case of a mismatch before billions of events get produced.

Reply:
Ok, following up on this: the model version is now set automatically with:

```
void l1t::AXOL1TLCondition::setModelVersion(const std::string modelversionname) {
  m_AXOL1TLmodelversion = "GTADModel_" + modelversionname;
  // LogDebug("AXOL1TLCondition") << "AXOL1TL model version: "<< m_AXOL1TLmodelversion << std::endl;                                                          
}
```

As per @makortel 's suggestion I also added the following when loading the model:

```
  try {
    model = loader.load_model();
  } catch(std::runtime_error& e) {
    LogDebug("AXOL1TLCondition") << "ERROR: failed to load model version "<< AXOL1TLmodelversion << ". Not evaluating condition!"<< std::endl;
    return false;
}
```

The result is that it still crashes when trying load_model, but now at least prints a message with the source of the error:

```
Cannot load library: asdahjsdksjd.so: cannot open shared object file: No such file or directory
ERROR: failed to load model version asdahjsdksjd. Not evaluating condition!
A fatal system signal has occurred: segmentation violation
```

I don't know of a way to avoid the crash purely within cms-sw without changing the cms-dist load_model dependancies, but please correct me if I'm wrong. As mentioned above, I think the way to fix this is to change load_model in cms-hls4ml/hls4mlEmulatorExtras included via cms-dist. [here](https://github.com/cms-hls4ml/hls4mlEmulatorExtras/blob/58f1ed7bf55a7ab74277035d83c419de7cbe28d3/src/hls4ml/emulator.cc#L16) and then make a PR to cms-dist to update it, but this is beyond the scope of this CMSSW PR. In any case, there is currently only one valid model version (v3) right now so this shouldn't cause any issues in the near future until this can be fixed. 

I created an issue in hls4mlemulatorextras accordingly: https://github.com/cms-hls4ml/hls4mlEmulatorExtras/issues/3

Let me know if this is an acceptable temporary solution for the scope of this PR.

Reply:
@quinnanm I have added a fix the model loader to be tested. Please see here: https://github.com/cms-hls4ml/hls4mlEmulatorExtras/releases/tag/v1.1.2 and here https://github.com/cms-hls4ml/hls4mlEmulatorExtras/pull/4.

PRs for this change to 14_1 and a backport to 14_0 are open here: https://github.com/cms-sw/cmsdist/pull/9064 and here: https://github.com/cms-sw/cmsdist/pull/9065

Reply:
model versions are now automatically set and cms exceptions behave as requested for unsupported models without returning a segfault